src/test/app/Batch_test.cpp

```diff
@@ -0,0 +1,2181 @@
+        for (auto const& signer : signers)
+        {
+            Serializer ss{
+                buildMultiSigningData(jtx::parse(ojv), signer.account.id())};
+            auto const sig = ripple::sign(
+                signer.account.pk(), signer.account.sk(), ss.slice());
+            jv[sfBatchSigners.jsonName][signer.index][sfBatchSigner.jsonName]
+              [sfAccount.jsonName] = signer.account.human();
+            jv[sfBatchSigners.jsonName][signer.index][sfBatchSigner.jsonName]
+              [sfSigningPubKey.jsonName] = strHex(signer.account.pk());
+            jv[sfBatchSigners.jsonName][signer.index][sfBatchSigner.jsonName]
+              [sfTxnSignature.jsonName] = strHex(Slice{sig.data(), sig.size()});
+        }
+        return jv;
+    }
+
+    XRPAmount
+    calcBatchFee(
+        test::jtx::Env const& env,
+        uint32_t const& signers,
+        uint32_t const& txns = 0)
+    {
+        XRPAmount const feeDrops = env.current()->fees().base;
+        return ((signers + 2) * feeDrops) + feeDrops * txns;
+    }
+
+    void
+    testEnable(FeatureBitset features)
+    {
+        testcase("enabled");
+
+        using namespace test::jtx;
+        using namespace std::literals;
+
+        for (bool const withBatch : {true, false})
+        {
+            auto const amend = withBatch ? features : features - featureBatch;
+            test::jtx::Env env{*this, envconfig(), amend};
+
+            auto const alice = Account("alice");
+            auto const bob = Account("bob");
+            auto const carol = Account("carol");
+            env.fund(XRP(1000), alice, bob, carol);
+            env.close();
+
+            auto const preAlice = env.balance(alice);
+            auto const preBob = env.balance(bob);
+
+            auto const seq = env.seq(alice);
+            auto const batchFee = calcBatchFee(env, 0, 1);
+
+            auto const txResult =
+                withBatch ? ter(tesSUCCESS) : ter(temDISABLED);
+
+            env(batch::batch(alice, seq, batchFee, tfAllOrNothing),
+                batch::add(pay(alice, bob, XRP(1)), seq + 1),
+                txResult);
+            env.close();
+        }
+    }
+
+    void
+    testPreflight(FeatureBitset features)
+    {
+        testcase("preflight");
+
+        using namespace test::jtx;
+        using namespace std::literals;
+
+        //----------------------------------------------------------------------
+        // preflight
+
+        test::jtx::Env env{*this, envconfig()};
+
+        auto const alice = Account("alice");
+        auto const bob = Account("bob");
+        auto const carol = Account("carol");
+        env.fund(XRP(1000), alice, bob, carol);
+        env.close();
+
+        // temINVALID_FLAG: Batch: invalid flags.
+        {
+            auto const seq = env.seq(alice);
+            auto const batchFee = calcBatchFee(env, 0, 0);
+            env(batch::batch(alice, seq, batchFee, tfAllOrNothing),
+                txflags(tfDisallowXRP),
```

Comment:
Can this not be simplified like this?
```suggestion
            env(batch::batch(alice, seq, batchFee, tfDisallowXRP),
```